### PR TITLE
Make sure individual form deletion populates deleted_on field

### PR DESCRIPTION
**Pull request by minhaminha**
## Technical Summary

When the ability to delete individual forms after archiving them was added, it called `form.save()` after the raw SQL function to update its values in the db was run, which overwrote those values to be null, except `state`, which was explicitly set on the model instance. This PR removes that line, changes the function name to be more accurate and adds a command to populate the `deleted_on` field on forms individually deleted in the last few months (to be picked up later by a periodic hard deletion task based on that value). 

### Safety story

Tested locally and on staging

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Comment by snopoke** [comment]
> When the ability to delete individual forms after archiving them was added, it attempted to run the soft_delete sql function from the model instance, which to my understanding prevented cursor.execute() from ultimately carrying out that function (updating the deleted_on, deletion_id and state values) since the cursor lacked the context it would've had if called from the model manger.

I don't follow this logic. What additional context would the cursor have when called from the manager?

Also, looking at the code in `soft_delete` it appears that it DOES call it from the manager: https://github.com/dimagi/commcare-hq/blob/dfa562f19b1c8af9c4b6bc52b736292436fd807c/corehq/form_processor/models/forms.py#L643

There are also 2 unit tests which verify this function:
* https://github.com/dimagi/commcare-hq/blob/f05c2485c3d0aa1b652ebc3272f942d9cb161974/corehq/form_processor/tests/test_basics.py#L86
* https://github.com/dimagi/commcare-hq/blob/33f38fc8b4424322a91a94fb0d62f1cd579b579e/corehq/form_processor/tests/test_kafka.py#L59

Perhaps you can create a test to illustrate the issue?

**Comment by snopoke** [comment]
Ah, I think I see the issue. The code is executing as follows:

```python
XFormInstance.objects.soft_delete_forms(form.domain, [form.form_id])
form.state |= self.DELETED
form.save()
```

So DB function updates sets the `deleted_on` date but then `form.save()` resets it back to null.

I think the resolution should be 2 things:
1. Update `soft_delete` to set the 'deleted_on' date to keep the model consistent with the DB
2. Don't call `.save()` since the DB update has already happened.

We should also update the tests and `form.is_deleted`.

**Comment by minhaminha** [comment]
> I don't follow this logic. What additional context would the cursor have when called from the manager?
> 
> Also, looking at the code in `soft_delete` it appears that it DOES call it from the manager:

Ah my apologies, I misunderstood/misapplied the docs concerning context managers to this scenario and mistakenly assumed that starting from the model instance (even if eventually called from the manager) would somehow limit the cursor's connection (??) but I realize now its all the same cursor object.

> So DB function updates sets the deleted_on date but then form.save() resets it back to null

I didn't even consider this but that would fully explain what was going on in that little block. Is it correct to say that this happens because `save()` copies the model values to the database (null, since it doesn't update on the model side)?

**Comment by snopoke** [comment]
> Is it correct to say that this happens because save() copies the model values to the database (null, since it doesn't update on the model side)?

Yes, that's correct.

**Comment by minhaminha** [comment]
> We should also update the tests and `form.is_deleted`.

did you mean removing the reference to `form.state` in `form.is_deleted`? This was done in [another PR](https://github.com/dimagi/commcare-hq/pull/33480).

File: corehq/form_processor/management/commands/populate_missing_delete_on_values.py
Comment (snopoke) [inline corehq/form_processor/management/commands/populate_missing_delete_on_values.py:14]: `xform.doc_type` is a computed property based on the presence of the `deleted_on` field so this won't work as a way to determine if the form 'should' be deleted.

However, since the bug also updated the `state` property, you should be able to use that to determine if the form is deleted or not.

Comment (minhaminha) [reply corehq/form_processor/management/commands/populate_missing_delete_on_values.py]: Thanks for pointing that out! I switched to using the old deleted state property along with Daniel's suggestion below

File: corehq/form_processor/management/commands/populate_missing_delete_on_values.py
Comment (minhaminha) [reply corehq/form_processor/management/commands/populate_missing_delete_on_values.py]: Thanks for the suggestion Daniel! I integrated both parts into the modified command

**Comment by minhaminha** [comment]
I'll do a trial run of the command on staging first, once I get approval on the modified version.

File: corehq/form_processor/management/commands/populate_missing_delete_on_values.py
Comment (millerdev) [inline corehq/form_processor/management/commands/populate_missing_delete_on_values.py:20]: Is it possible to include the _deleted_ state filter here? (`state & 64 = 64`) That would greatly reduce the number of records returned by this iteration.

Comment (minhaminha) [reply corehq/form_processor/management/commands/populate_missing_delete_on_values.py]: added the extra filter!

File: corehq/form_processor/management/commands/populate_missing_delete_on_values.py
Comment (millerdev) [inline corehq/form_processor/management/commands/populate_missing_delete_on_values.py:20]: nit
```suggestion
                Q(deleted_on__isnull=True, state__gte=old_deleted_state),
```

**Review by millerdev — APPROVED** [review]

**Comment by minhaminha** [comment]
After testing and trying to debug this on staging, I can't seem to make the `deleted_on` population command run as intended. The initial query/chunking is returning the right limited sets of forms but the raw SQL function doesn't seem to be getting applied. Is there something else I'd forgotten to consider when running this on staging/prod - particularly having to do with partitioned data that might make this ineffective as opposed to running it locally? The only thing I can think to do now is set the `delete_on` value in the code but at the cost of performance efficiency (though the impact might actually not be that much with the chunking).

File: corehq/form_processor/management/commands/populate_missing_delete_on_values.py
Comment (snopoke) [inline corehq/form_processor/management/commands/populate_missing_delete_on_values.py:29]: This cursor is pointing to the 'proxy' database which has no data on it. That's why this isn't working on staging. You'll need to use `split_list_by_db_partition` to split the list of IDs by database and then do a query per DB:

```python
for db, db_chunk in split_list_by_db_partition(chunk):
    with XFormInstance.get_cursor_for_partition_db(db) as cursor:
        ...
```

Comment (minhaminha) [reply corehq/form_processor/management/commands/populate_missing_delete_on_values.py:29]: ahhh yea that'll do it. Now I'm curious though - what is the purpose the proxy database? Is it just for testing?

Comment (snopoke) [reply corehq/form_processor/management/commands/populate_missing_delete_on_values.py]: If you are calling one of the PlProxy functions you need to do that on the proxy database

**Comment by minhaminha** [comment]
Just confirmed this works on staging!

**Pull request merged**
